src/utils/offlineRequestHandler.ts

```diff
@@ -0,0 +1,415 @@
+    Authorization?: string;
+  };
+  method: string;
+  body: any;
+  keepalive?: boolean;
+}
+
+interface SerializedRequestBody {
+  type: 'blob' | 'formData' | 'arrayBuffer' | 'text';
+  body: ArrayBuffer | Record<string, any> | string;
+}
+
+interface SerializedRequest {
+  formKey: string;
+  url: string;
+  method: string;
+  headers: string;
+  body: SerializedRequestBody['body'];
+  bodyType: SerializedRequestBody['type'];
+  timestamp: number;
+  type: string;
+  stepKey?: string;
+  keepalive?: boolean;
+}
+
+const beforeUnloadEventHandler = (event: any) => {
+  // Recommended
+  event.preventDefault();
+
+  // Included for legacy support, e.g. Chrome/Edge < 119
+  event.returnValue = true;
+};
+
+let unloadCounter = 0;
+
+const trackUnload = () => {
+  if (unloadCounter === 0)
+    featheryWindow().addEventListener('beforeunload', beforeUnloadEventHandler);
+  unloadCounter++;
+};
+
+const untrackUnload = () => {
+  unloadCounter--;
+  if (unloadCounter === 0)
+    featheryWindow().removeEventListener(
+      'beforeunload',
+      beforeUnloadEventHandler
+    );
+};
+
+export function useOfflineRequestHandler(formKey: string) {
+  useEffect(() => {
+    if (!runningInClient()) return;
+
+    offlineRequestHandler.replayRequests(formKey);
+    const handleOnline = () => offlineRequestHandler.replayRequests(formKey);
+    featheryWindow().addEventListener('online', handleOnline);
+    return () => featheryWindow().removeEventListener('online', handleOnline);
+  }, []);
+}
+
+export class OfflineRequestHandler {
+  private isReplayingRequests: Map<string, boolean>;
+  private readonly dbName: string;
+  private readonly storeName: string;
+  private readonly dbVersion: number;
+  private readonly maxRetryAttempts: number;
+  private readonly retryDelayMs: number;
+  private onlineSignals: Map<string, any[]>;
+  private indexedDBSupported: boolean;
+
+  constructor(
+    dbName: string,
+    storeName: string,
+    dbVersion: number,
+    maxRetryAttempts: number,
+    retryDelayMs: number
+  ) {
+    this.isReplayingRequests = new Map();
+    this.dbName = dbName;
+    this.storeName = storeName;
+    this.dbVersion = dbVersion;
+    this.maxRetryAttempts = maxRetryAttempts;
+    this.retryDelayMs = retryDelayMs;
+    this.onlineSignals = new Map();
+    this.indexedDBSupported = typeof indexedDB !== 'undefined';
+  }
+
+  // Check if any requests are stored in indexedDB
+  public async dbHasRequest(): Promise<boolean> {
+    const { store } = await this.getDbTransaction('readonly');
+    const count = await new Promise<number>((resolve) => {
+      const request = store.count();
+      request.onsuccess = () => resolve(request.result);
+    });
+
+    return count > 0;
+  }
+
+  public onlineAndReplayed(formKey: string) {
+    return new Promise((resolve) => {
+      if (!this.onlineSignals.has(formKey)) {
+        this.onlineSignals.set(formKey, []);
+      }
+      this.onlineSignals.get(formKey)!.push(resolve);
```

Comment:
what is the exclamation mark here for?

Reply:
It's used for Typescript to assert that the expression before it is not null or undefined